#! /bin/sh -e

# DP: - When running the i386 binaries on amd64, adjust the properties
# DP:   java.home, gnu.classpath.home.url, sun.boot.class.path,
# DP:   gnu.gcj.precompiled.db.path.

dir=
if [ $# -eq 3 -a "$2" = '-d' ]; then
    pdir="-d $3"
    dir="$3/"
elif [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch)
        patch $pdir -f --no-backup-if-mismatch -p0 < $0
	cd ${dir}libjava && autoconf
        ;;
    -unpatch)
        patch $pdir -f --no-backup-if-mismatch -R -p0 < $0
	rm -f ${dir}libjava/configure
        ;;
    *)
        echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
        exit 1
esac
exit 0

--- libjava/gnu/classpath/natSystemProperties.cc.orig	2006-09-10 14:07:47.130718000 +0200
+++ libjava/gnu/classpath/natSystemProperties.cc	2006-09-10 16:30:31.640718000 +0200
@@ -209,6 +209,18 @@
   // part we do this because most people specify only --prefix and
   // nothing else when installing gcj.  Plus, people are free to
   // redefine `java.home' with `-D' if necessary.
+  struct utsname u;
+#if defined(__linux__) && defined (__i386__)
+  uname (&u);
+  if (! strcmp ("x86_64", u.machine))
+    {
+      SET ("java.home", "/usr/lib32/jvm/java-1.5.0-gcj-4.1-1.5.0.0/jre");
+      SET ("gnu.classpath.home", "/usr");
+      SET ("gnu.classpath.home.url", "file:///usr/lib/../lib32");
+    }
+  else
+#endif
+    {
   SET ("java.home", JAVA_HOME);
   SET ("gnu.classpath.home", PREFIX);
   // This is set to $(toolexeclibdir) because we use this to find
@@ -217,11 +229,11 @@
   strcpy (val2, "file://");
   strcat (val2, TOOLEXECLIBDIR);
   SET ("gnu.classpath.home.url", val2);
+    }
 
   SET ("file.encoding", default_file_encoding);
 
 #ifdef HAVE_UNAME
-  struct utsname u;
   if (! uname (&u))
     {
       SET ("os.name", u.sysname);
@@ -348,11 +360,21 @@
   // without overriding java.endorsed.dirs.
   SET ("gnu.gcj.runtime.endorsed.dirs", GCJ_ENDORSED_DIRS);
 
+#if defined(__linux__) && defined (__i386__)
+  if (! strcmp ("x86_64", u.machine))
+    {
+      SET ("sun.boot.class.path", "/usr/lib32/jvm/java-1.5.0-gcj-4.1-1.5.0.0/jre/lib/rt.jar");
+      SET ("gnu.gcj.precompiled.db.path", "/var/lib32/gcj-4.1/classmap.db");
+    }
+  else
+#endif
+    {
   // The path to libgcj's boot classes
   SET ("sun.boot.class.path", BOOT_CLASS_PATH);
 
   // If there is a default system database, set it.
   SET ("gnu.gcj.precompiled.db.path", LIBGCJ_DEFAULT_DATABASE);
+    }
 
   // Set some properties according to whatever was compiled in with
   // `-D'.  Important: after this point, the only properties that
